src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -1547,5 +1622,40 @@ internal override SyntaxNode ScopeDesignator
                 return _syntax;
             }
         }
+
+        private MethodArgumentInfo GetSpecialTypeMemberInfo(SpecialMember member, SyntaxNode syntax, DiagnosticBag diagnostics)
+        {
+            return (MethodSymbol)GetSpecialTypeMember(member, diagnostics, syntax) is { } resolvedMember
+                    ? MethodArgumentInfo.ParameterlessMethod(resolvedMember)
+                    : null;
+        }
+
+        /// <param name="extensionReceiverOpt">If extension methods are allowed, this must be non-null.</param>
+        private MethodArgumentInfo BindDefaultArguments(MethodSymbol method, BoundExpression extensionReceiverOpt, bool allowExtensionMethods, bool expanded, SyntaxNode syntax, DiagnosticBag diagnostics)
+        {
+            Debug.Assert(!method.IsExtensionMethod || allowExtensionMethods);
+            Debug.Assert(extensionReceiverOpt != null || !allowExtensionMethods);
```

Comment:
Can this assertion be strengthened? `(extensionReceiverOpt != null) == allowExtensionMethods`